Visualising waypoints3

forcecore opened the pull request:
In continuation of https://github.com/OpenRA/OpenRA/pull/11998 ... for issue #2598

This should almost feel like RA2 except that the current waypoint system doesn't allow patrolling. EA's C&C3 and onwards didn't have that anyway.

https://youtu.be/OOavDMVlqUY

Turupawn committed:
visualising waypoints (round 2)

Turupawn committed:
small fixes

Turupawn committed:
getTargets() refactor, it's still glitchy and messy but I think it's heading to a good direction

Turupawn committed:
getTargets() now returning an Enumerable

Turupawn committed:
wiped SetTargetLine()

Turupawn committed:
fix: check pass

Turupawn committed:
Drawing the last valid color

forcecore committed:
Updated to build on current bleed

forcecore committed:
Waypoint works almost like RA2

pchote commented on OpenRA.Mods.Common/Activities/Move/Follow.cs:
```diff
@@ -47,5 +49,14 @@ public override Activity Tick(Actor self)
 		}
+
+		// Follow, is as in follow and attack (chase) so it is actually an attack activity.
+		// (What came to my mind first was follow and guard activity)
```
This is used for attacking *and* guarding.  The guard case should return yellow.

forcecore replied:
Fixed

pchote commented on OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs:
```diff
@@ -153,12 +153,19 @@ List<CPos> CalculatePathToTarget(Actor self)
-			return Target.None;
+		public override IEnumerable<Target> GetTargets(Actor self)
+		{
+			yield return Target;
```
You removed the `inner.GetTargets(self)` above but didn't replace it here.  Is this intentional?

forcecore replied:
It was removed for simplification. What ever it does in the inner loop, it is moving towards target and the next activity is more important. Landing sequence for example, I wanted to simplify it and prevent moving to the left of the airfield before Land activity to be not drawn but I couldn't do that neatly.

forcecore committed:
Two wrong colorization fixed

Follow color: now draws follow-guard as yellow
Force fire after move addressed (wasn't even drawn)

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -261,18 +262,24 @@ void SanityCheckPath(Mobile mobile)
 		}
 
-		abstract class MovePart : Activity
+		public abstract class MovePart : Activity
```
Carried over from #11998/#11794:  This should stay private.

forcecore replied:
In relation to below struct approach, this gives rise to make struct a linked list, as this MovePart activity is quite different from other activities. It doesn't have NextActivity nor ParentActivity but in practice, the member variable Move is the next activity. Then the below struct must become a class.

Turupawn replied:
@forcecore Indeed, I made MovePart public in order to access it from the [RenderAboveShroud()](https://github.com/forcecore/OpenRA/blob/9948bb44dd0fa7852a1b57897a21b0f5d742d261/OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs#L70) method.

Great to hear this is beeing continued! :dolphin:

pchote commented on OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs:
```diff
@@ -62,80 +44,62 @@ void INotifySelected.Selected(Actor a)
+					prev = target.CenterPosition;
+				}
+
+				if (a is ResupplyAircraft)
```
Hardcoding activities here is no good because it prevents any downstream modders from implementing their own "terminal" activities.

pchote replied:
The best solution here is to replace the `TargetLineColor` / `GetTargets` with the a new metadata struct that specifies for each segment the target, color, and whether the line rendering should stop after that node:

```csharp
public struct TargetLineNode
{
	public readonly Target Target;
	public readonly Color Color;
	public readonly bool Terminal;
	public TargetLineStep(Target target, Color color, bool terminal=false) { /* set the fields */ }
}
```

The target / color query split was already an issue I wasn't happy with (discussed at length during the earlier revisions of this feature), so this kills two birds with one stone.

forcecore replied:
I'm uncertain I can make struct approach to work.
(1) Attack move activity changes target
(2) Color depends on the next activity, which dynamically change too as the queue can be appended later on by player's shift clicks (same goes for Target)

pchote replied:
(1) shouldn't be a problem, you query these once, render them, (possibly persist them until the next tick), and then discard / recalculate them next time.

For (2) query them, ToList() the enumerable, and then take the color from the last entry.

forcecore replied:
I see, you want to use TargetLineNode as a return value so that everything needed is returned in one shot.

pchote commented on OpenRA.Game/Activities/Activity.cs:
```diff
@@ -293,12 +296,14 @@ public override Activity NextActivity
 
 	public static class ActivityExts
 	{
+/*
```
Dead code should be removed, not just commented.

Mailaender commented on OpenRA.Mods.Common/Activities/Air/Fly.cs:
```diff
@@ -10,6 +10,7 @@
 #endregion
 
 using System.Collections.Generic;
+using System.Drawing;
```
This change can be omitted I assume.

Mailaender commented on OpenRA.Mods.Common/Activities/Air/HeliFly.cs:
```diff
@@ -10,6 +10,7 @@
 #endregion
 
 using System.Collections.Generic;
+using System.Drawing;
```
Same here.

Mailaender commented on OpenRA.Mods.Common/Activities/DonateCash.cs:
```diff
@@ -9,6 +9,8 @@
 #endregion
 
+using System.Collections.Generic;
+using System.Drawing;
```
and here.

forcecore committed:
Minor fixups

Removed unneeded using directive.
Removed a completely commented function.

forcecore committed:
TargetLineNode introduced

No more TargetLineColor.

forcecore commented:
TargetLineColor struct is now returned

forcecore committed:
class MovePart is private again.

pchote commented on OpenRA.Game/Activities/Activity.cs:
```diff
@@ -266,6 +282,11 @@ public virtual IEnumerable<Target> GetTargets(Actor self)
 			yield break;
 		}
+
+		public virtual TargetLineNode TargetLineNode(Actor self)
```
Please keep this as an `IEnumerable<TargetLineNode>`. Composite activities may need to return more than one node.

forcecore replied:
Can you give me an example? All compositie activities gets "unrolled" and I had no trouble cycling through all of them with Next.

pchote commented on OpenRA.Game/Activities/Activity.cs:
```diff
@@ -11,13 +11,29 @@
+
+		public TargetLineNode(Target target, Color color, Activity next)
+		{
+			// Note: Not all activities are drawable. In that case, pass Target.Invalid as target.
```
These activities should `yield break` to instead not return any nodes.

pchote commented on OpenRA.Mods.Common/Activities/Air/Fly.cs:
```diff
@@ -88,6 +89,12 @@ public override IEnumerable<Target> GetTargets(Actor self)
+
+		public override TargetLineNode TargetLineNode(Actor self)
+		{
+			var color = NextActivity == null ? Color.Green : NextActivity.TargetLineNode(self).Color;
```
This is going to hurt performance and maintainability.  Each target line needs to return the properties of *itself*, without depending on what the next segment is doing.  When you draw the lines you can then look at the properties of the last node to get the color to draw the whole chain.

forcecore replied:
Looking at the last node isn't correct. If I issue an attack order then make the unit afterwards, everything will be drawn in green. That's why I didn't get your struct approach. Two solutions, I think.

(1) Giveup color altogether during waypoint mode and paint everything blue. (2) Revert to color property in older commit

obrakmann replied:
Shouldn't each segment be drawn in the colour that the respective activity specifies?

forcecore replied:
I wish I could do that, it would be much easier. However, even Attack order involves "moving to the target" and makes most part visualization drawing in green. My guess is that if ALL activities can be refactored into parent-child activity and, if all I have to do becomes "traverse NextInQueue" then the color can be statically determined.

obrakmann commented on OpenRA.Mods.Cnc/Traits/Infiltration/Infiltrates.cs:
```diff
@@ -119,7 +119,7 @@ public void ResolveOrder(Actor self, Order order)
 				self.CancelActivity();
 
-			self.SetTargetLine(target, Color.Red);
+			self.ShowTargetLines();
```
I think it's wrong that this is called in trait code at all. It just causes the lines to appear at the wrong time when you queue commands. It should instead probably be called somewhere in `Activity.OuterTick`, although then you have to figure out how to deal with the engine/mod code separation.

forcecore replied:
Unless SetTargetLine's display parameter was explicitly specified as false, the game did show the target line. In Infiltrates.cs' case, this is correct. In some cases I put ShowTargetLines is false. I'll fix that

forcecore replied:
Fixed.

pchote replied:
Can we do this from the same place that we show the target/move flash?  That seems like it would be the most natural place to me.

forcecore commented:
Actually, I have a new neater idea, utilizing SetTargetLine. I'll start from scratch and make a new PR.

forcecore commented on OpenRA.Game/Activities/Activity.cs:
```diff
@@ -11,13 +11,29 @@
+	{
+		public readonly Target Target;
+		public readonly Color Color;
+		public readonly Activity Next;
```
I really want to use it but MovePart returns null for NextActivity

forcecore replied:
I really really want to use it but some activities like Move has its own non-standard way of managing "next" activity that RootActivity, NextInQueue, NextActivity are all null.

forcecore committed:
Targeting actor didn't flash the frozen proxy

forcecore committed:
Removed SetTargetLine() where display param was false.

forcecore committed:
TargetLineNode is returned  without refering to NextActivity

forcecore committed:
Refactored Move.cs to use Parent-Child activity

forcecore committed:
Polished to look good

forcecore commented:
Two activities have been refactored to remove ambiguity when drawing waypoints.
GuardTargetActivity.cs : New activity to wrap Follow-Guard case.
Move.cs : Instead of only using Move activity in MovePart and the derivatives, Move and MovePart are now parent-child activities.

obrakmann commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -185,8 +185,9 @@ public override Activity Tick(Actor self)
 					mobile.Facing,
 					mobile.Facing,
 					0);
+				ChildActivity = move;
```
You can assign directly to `ChildActivity` here, no need for the intermediate variable.

There's another line further up  that would need to change. Unfortunately I can't put a comment there.
167: `return ActivityUtils.SequenceActivities(new Turn(self, firstFacing), this);`

Reusing activities like that is a bug. It would be better to queue the `Turn` as a child instead and then run that.

obrakmann commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -185,8 +185,9 @@ public override Activity Tick(Actor self)
+				ChildActivity = move;
 
-				return move;
+				return ChildActivity;
```
This line goes against the whole idea of child activities, and is also the reason why the change to `MovePart.Queue` below is needed in the first place. You should be able to use `ActivityUtils.RunActivity(self, ChildActivity)` somewhere in this `Tick` method instead. Changing that will also prevent the `MovePart`s from becoming the actor's `CurrentActivity`.

obrakmann commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -273,9 +274,7 @@ public override IEnumerable<Target> GetTargets(Actor self)
-			//var color = NextActivity == null ? Color.Green : NextActivity.TargetLineNode(self).Color;
-			//return new TargetLineNode(Target.FromCell(self.World, this.destination.Value), color, NextActivity);
-			return new TargetLineNode(Target.FromCell(self.World, this.destination.Value), Color.Aqua, false);
+			return new TargetLineNode(Target.FromCell(self.World, this.destination.Value), Color.Green, false);
```
Is this line supposed to be in this commit?

obrakmann commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -326,13 +325,15 @@ public MovePart(Move move, WPos from, WPos to, int fromFacing, int toFacing, int
 			public override void Queue(Activity activity)
 			{
-				Move.Queue(activity);
+				ParentActivity.Queue(activity);
```
see comment above wrt to the `return ChildActivity` line.

Btw, the `readonly Move move` field can either be removed, or be assigend to `ParentActivity as Move` in the ctor, and then be used instead of ParentActivity. You'd avoid all the casts below that way.

obrakmann commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -326,13 +325,15 @@ public MovePart(Move move, WPos from, WPos to, int fromFacing, int toFacing, int
+				// MovePart has no child or queued activity so it is correct to cancel this
+				// then return the RootActivity's cancel result.
+				base.Cancel(self);
+				return RootActivity.Cancel(self, keepQueue);
```
I don't think it is correct to cancel the whole queue. You don't know from which level of the tree the `Cancel` originates from. It could be from an activity that decides it wants to do something else instead of moving, but still continue running itself. This line would cancel that activity.

obrakmann reviewed (changes requested):
> Move.cs : Instead of only using Move activity in MovePart and the derivatives, Move and MovePart are now parent-child activities.

Interesting. I agree that refactoring `Move` is absolutely necessary before any more work on activities can be done. It's quite a roadblock at the moment, and that's why I have tried my hand at refactoring `Move` myself, see https://github.com/OpenRA/OpenRA/compare/bleed...obrakmann:activities-pt3. My version is still buggy, though, while yours does not quite finish the conversion. I have addressed the issues with your version inline. I hope that between your version and mine, we can produce one that both works and is a complete conversion.

pchote commented:
Can we please split the move refactoring out to its own PR?

forcecore commented:
Rather than that I can wait until at least Move is done, rather than bringing conflicts. Do you like the direction though?

forcecore commented:
@obrakmann In the next playtest, waypoint is exposed to UI so this might become high priority. What if I refactor Move in another PR and come back to this? The refactor might not be what obrakmann is thinking but he can do is work on it I guess?

reaperrr commented:
I don't think this should block the playtest (though it would be better if it made it in), but I agree it's a must for next release.

reaperrr commented:
Needs rebase as well.

pchote commented:
I consider this to be one of our high priorities for Next + 1, now that we have branched `prep-1707`.  @forcecore would you be able to split off a new PR for the Move stuff discussed above?

forcecore commented:
Now that I've made my huge dock PR, I'll get my hands on this one.

pchote commented:
I expect it will take a while for the prerequisites to be merge (#13995 has been sitting unreviewed for over two weeks, and its only the very first step!), so moving to the Future milestone to help clean up the PR queue.
